### DISCUSSION TITLE
Is it possible to have Collections.IsNullEmpty()/Enumerable.IsNullOrEmpty() in framework

### DISCUSSION THREAD

**Discussion by shinkarmv** [dotnet/runtime#43639] [General] [answered]
With current available functions to check null/ empty collections/enumerable

we need to write some extension method or add explicit check on collection/enumerable for null and count ==0.

Also to check empty - we can apply Any method from Linq but it throws exception null.

So is it possible to have Collections.IsNullEmpty()/Enumerable.IsNullOrEmpty() in framework.

Thoughts?

**Comment by theunrepentantgeek** [comment]
This would be more properly a request for new methods in the base class library, not a change to the C# language.

However, given the strong push to detect **null** values and enable programmers to avoid having **null** show up unexpectedly, I'm not sure they would gain much traction.

While explicit handling is required, the current solutions are pretty concise:

```
// Using Any()
IEnumerable<Foo>? myEnumerable = ... // elided
if (myEnumerable?.Any() ?? false) 
{
    // myEnumerable has items
}

// Using Count
IList<Foo>? myList = ... // elided
if ((myList?.Count ?? 0) > 0) 
{
    // myList has items
}

// Custom Extension method 
if (myEnumerable.HasItems() > 0) 
{
    // myEnumerable has items
}
```

FWIW, I used the name `HasItems()` because **true** would imply that you have a sequence _and_ it has items. An `IsEmptyOrNull()` method is ambiguous when it returns **true** - is it **null**, or is it an empty sequence?

**Reply by svick** [reply]
Personally, I think these forms are more readable:

```c#
if (myEnumerable?.Any() == true) 
if (myList?.Count > 0) 
```

But that doesn't change anything about your point.

**Reply by alrz** [reply]
With https://github.com/dotnet/csharplang/issues/3435 you could do `expr is [_, ..]`  ;)

However, the advantage of an extension method is that you can optimize for IList at runtime.

**Reply by shinkarmv** [reply]
@theunrepentantgeek - Thank you the response and proposed more feasible function to handle this.

So are we going to add HasItems() to Base Class Library?

Regards,
Mukesh

**Reply by theunrepentantgeek** [reply]
That would more properly be a request for the base class library repo, not for here. Log an issue over at [dotnet/runtime](https://github.com/dotnet/runtime). If they're interested, they'll likely accept a PR.

**Reply by YairHalberstadt** [reply]
@theunrepentantgeek this discussion has been moved to dotnet/runtime.

**Reply by theunrepentantgeek** [reply]
Oh! I didn't notice that. #embarrassing 😊

**Comment by NN---** [comment]
While it is convenient , there is much work defining how it should be implemented.
It is actually very easy to implement own extension method.
And there are plenty libraries who do it for instance:
https://github.com/rsdn/CodeJam/wiki/M_CodeJam_Collections_CollectionExtensions_IsNullOrEmpty__1

**Comment by danmoseley** [comment, marked as answer]
Good answers above in comments. Can't mark comment as answer. Marking as answered.
